src/io/gml.c

```diff
@@ -319,65 +339,89 @@ igraph_error_t igraph_read_graph_gml(igraph_t *graph, FILE *instream) {
                     IGRAPH_FINALLY_CLEAN(2);
                 } else {
                     /* already seen, should we update type? */
                     igraph_attribute_record_t *atrec = VECTOR(vattrs)[trieid];
                     igraph_attribute_type_t type1 = atrec->type;
-                    igraph_i_gml_tree_type_t type2 = igraph_gml_tree_type(node, j);
-                    if (type1 == IGRAPH_ATTRIBUTE_NUMERIC && type2 == IGRAPH_I_GML_TREE_STRING) {
+                    if (type == IGRAPH_I_GML_TREE_STRING) {
                         atrec->type = IGRAPH_ATTRIBUTE_STRING;
+                    } else if (type1 == -1) {
+                        if (type == IGRAPH_I_GML_TREE_INTEGER || type == IGRAPH_I_GML_TREE_REAL) {
+                            atrec->type = IGRAPH_ATTRIBUTE_NUMERIC;
+                        }
                     }
                 }
                 /* check id */
-                if (!hasid && !strcmp(name, "id")) {
-                    igraph_integer_t id;
+                if (!strcmp(name, "id")) {
+                    igraph_integer_t id, trie_id;
+                    igraph_integer_t trie_size =igraph_trie_size(&trie);
+                    if (hasid) {
+                        /* A 'node' must not have more than one 'id' field. */
+                        IGRAPH_ERRORF("Node has multiple id fields in GML file, line %" IGRAPH_PRId ".", IGRAPH_PARSEERROR,
+                                      igraph_gml_tree_line(node, j));
+                    }
                     if (igraph_gml_tree_type(node, j) != IGRAPH_I_GML_TREE_INTEGER) {
-                        IGRAPH_ERROR("Non-integer node id in GML file.", IGRAPH_PARSEERROR);
+                        IGRAPH_ERRORF("Non-integer node id in GML file, line %" IGRAPH_PRId ".", IGRAPH_PARSEERROR,
+                                      igraph_gml_tree_line(node, j));
                     }
                     id = igraph_gml_tree_get_integer(node, j);
-                    snprintf(cname, sizeof(cname) / sizeof(char) -1, "%" IGRAPH_PRId, id);
-                    IGRAPH_CHECK(igraph_trie_get(&trie, cname, &id));
+                    snprintf(cname, sizeof(cname) / sizeof(char) - 1, "%" IGRAPH_PRId, id);
+                    IGRAPH_CHECK(igraph_trie_get(&trie, cname, &trie_id));
+                    if (trie_id != trie_size) {
+                        /* This id has already been seen in a previous node. */
+                        IGRAPH_ERRORF("Duplicate node id in GML file, line %" IGRAPH_PRId ".", IGRAPH_PARSEERROR,
+                                      igraph_gml_tree_line(node, j));
+                    }
                     hasid = 1;
                 }
             }
             if (!hasid) {
-                IGRAPH_ERROR("Node without 'id' while parsing GML file.", IGRAPH_PARSEERROR);
+                /* Isolated nodes are allowed not to have an id.
+                 * We generate an "n"-prefixed string id to be used in the trie. */
```

Comment:
Let's try to do something less predictable; I can imagine a graph where the first node is named `n1` and the next node is unnamed so it would also get `n1` as its name.

Ideas:

* Use `_generated_id_%d` or something similar that clearly indicates that this is a generated ID and that is less likely to be used in normal circumstances.
* Stick with `n%d` but then make sure to pick the first unused ID. If the generated ID is already used, increase the number until we obtain a name that is not used.

Reply:
These generated IDs are never exposed to the user, and can never conflict. These are just used in the trie that translates GML ids to igraph vertex IDs.

Normally, GML ids can be any integer, while igraph vertex IDs must be consecutive integers. They way the GML reader works is that it converts the GML id to a string, and inserts it into a trie. Then the index of that string in the trie will be the igraph vertex ID. By adding the `n` prefix, we can ensure that these "generated string ids" never conflict with a explicit GML id. By keeping the prefix short (a single letter), we allow the trie to be more efficient for both storage and lookup.

Reply:
Ah, okay, so there is no way for the user to create a node with ID equal to `n1`? That's okay then.

Reply:
This is explained in the comments here:

https://github.com/igraph/igraph/blob/3dd467fcc074db9e8d5f54ef8566e9f5c66a8955/src/io/gml.c#L298